File: c++/classes/inheritance/base_class_elements_access.cpp
#include <iostream>

using std::string, std::cout, std::endl, std::boolalpha, std::noboolalpha;

class Mammal {
public:
  bool has_tail;
  static bool is_milk_feeded;
  const string class_taxon = "Mammalia";

  void display() {
    cout << "Class: " << class_taxon << endl
      << "Class: " << this->class_taxon << endl
      << "Class: " << Mammal::class_taxon << endl
      << boolalpha
      << "Is it milk-feeded as a cub? " << is_milk_feeded << endl
      << "Is it milk-feeded as a cub? " << this->is_milk_feeded << endl
      << "Is it milk-feeded as a cub? " << Mammal::is_milk_feeded << endl
      << "Does it have tail? " << has_tail << endl
      << "Does it have tail? " << this->has_tail << endl
      << "Does it have tail? " << Mammal::has_tail << endl
      << noboolalpha;
  }
};

bool Mammal::is_milk_feeded = true;

class Fox : public Mammal {
public:
  string name;
  static bool has_fur;
  const string species_taxon = "Vulpes vulpes";

  void show() {
    cout << "Hi, my name is " << name << "." << endl
      << "Hi, my name is " << this->name << "." << endl
      << "Hi, my name is " << Fox::name << "." << endl
      << "Class: " << class_taxon << endl
      << "Class: " << this->class_taxon << endl
      << "Class: " << Mammal::class_taxon << endl
      << "Class: " << Fox::class_taxon << endl
      << "Species: " << species_taxon << endl
      << "Species: " << this->species_taxon << endl
      << "Species: " << Fox::species_taxon << endl
      << boolalpha
      << "Am I milk-feeded as a cub? " << is_milk_feeded << endl
      << "Am I milk-feeded as a cub? " << this->is_milk_feeded << endl
      << "Am I milk-feeded as a cub? " << Mammal::is_milk_feeded << endl
      << "Am I milk-feeded as a cub? " << Fox::is_milk_feeded << endl
      << "Do I have tail? " << has_tail << endl
      << "Do I have tail? " << this->has_tail << endl
      << "Do I have tail? " << Mammal::has_tail << endl
      << "Do I have tail? " << Fox::has_tail << endl
      << "Do I have fur? " << has_fur << endl
      << "Do I have fur? " << this->has_fur << endl
      << "Do I have fur? " << Fox::has_fur << endl
      << noboolalpha;
  }

  void display() {
    cout << "Name: " << name << "." << endl
      << "Name: " << this->name << "." << endl
      << "Name: " << Fox::name << "." << endl
      << "Species: " << species_taxon << endl
      << "Species: " << this->species_taxon << endl
      << "Species: " << Fox::species_taxon << endl;
      Mammal::display();
      cout << boolalpha
      << "Does it have fur? " << has_fur << endl
      << "Does it have fur? " << this->has_fur << endl
      << "Does it have fur? " << Fox::has_fur << endl
      << noboolalpha;
  }
};

bool Fox::has_fur = true;

int main() {
  Fox fox_ferdinand;
  fox_ferdinand.name = "Ferdinand";

  fox_ferdinand.show();
  cout << endl;
  fox_ferdinand.display();
  cout << endl;
}

File: pascal/classes/inheritance/base_class_elements_access.pas
{$MODE OBJFPC}

program BaseClassElementsAccess;

type
  Mammal = class
  public
    has_tail: boolean;
    is_milk_feeded: boolean; static;
  const
    class_taxon = 'Mammal';
  public
    procedure display();
    constructor create();
  end;

  Fox = class(Mammal)
  public
    name: string;
    has_fur: boolean; static;
  const
    species_taxon = 'Vulpes vulpes';
  public
    procedure show();
    procedure display();
    constructor create();
  end;

constructor Mammal.create();
begin
  is_milk_feeded := true;
end;

procedure Mammal.display();
begin
  writeln('Class: ', class_taxon);
  writeln('Class: ', Self.class_taxon);
  writeln('Class: ', Mammal.class_taxon);
  writeln('Is it milk-feeded as a cub? ', is_milk_feeded);
  writeln('Is it milk-feeded as a cub? ', Self.is_milk_feeded);
  writeln('Is it milk-feeded as a cub? ', Mammal.is_milk_feeded);
  writeln('Does it have tail? ', has_tail);
  writeln('Does it have tail? ', Self.has_tail);
end;

constructor Fox.create();
begin
  Mammal.create();
  has_fur := true;
end;

procedure Fox.show();
begin
  writeln('Hi, my name is ', name, '.');
  writeln('Hi, my name is ', Self.name, '.');
  writeln('Class: ', class_taxon);
  writeln('Class: ', Self.class_taxon);
  writeln('Class: ', Mammal.class_taxon);
  writeln('Class: ', Fox.class_taxon);
  writeln('Species: ', species_taxon);
  writeln('Species: ', Self.species_taxon);
  writeln('Species: ', Fox.species_taxon);
  writeln('Am I milk-feeded as a cub? ', is_milk_feeded);
  writeln('Am I milk-feeded as a cub? ', Self.is_milk_feeded);
  writeln('Am I milk-feeded as a cub? ', Mammal.is_milk_feeded);
  writeln('Am I milk-feeded as a cub? ', Fox.is_milk_feeded);
  writeln('Do I have tail? ', has_tail);
  writeln('Do I have tail? ', Self.has_tail);
  writeln('Do I have fur? ', has_fur);
  writeln('Do I have fur? ', Self.has_fur);
  writeln('Do I have fur? ', Fox.has_fur);
  writeln('Do I have fur? ', Self.has_fur);
end;

procedure Fox.display();
begin
  writeln('Name: ', name, '.');
  writeln('Name: ', Self.name, '.');
  writeln('Species: ', species_taxon);
  writeln('Species: ', Self.species_taxon);
  writeln('Species: ', Fox.species_taxon);
  Mammal(Self).display();
  writeln('Does it have fur? ', has_fur);
  writeln('Does it have fur? ', Self.has_fur);
  writeln('Does it have fur? ', Fox.has_fur);
  writeln('Does it have fur? ', Self.has_fur);
end;

var
  fox_ferdinand: Fox;

begin
  fox_ferdinand := Fox.create();
  fox_ferdinand.name := 'Ferdinand';

  fox_ferdinand.display();
  writeln();
  fox_ferdinand.show();
  writeln();
end.
